libphal/phal_sbedump.C

```diff
@@ -0,0 +1,228 @@
+extern "C" {
+#include <libgen.h>
+#include <libpdbg.h>
+#include <libpdbg_sbe.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <unistd.h>
+}
+
+#include "libphal.H"
+#include "log.H"
+
+#include <ekb/chips/p10/procedures/hwp/lib/p10_pibmem_dump.H>
+#include <ekb/chips/p10/procedures/hwp/lib/p10_pibms_reg_dump.H>
+#include <ekb/chips/p10/procedures/hwp/lib/p10_ppe_state.H>
+#include <ekb/chips/p10/procedures/hwp/lib/p10_sbe_localreg_dump.H>
+#include <ekb/chips/p10/procedures/hwp/lib/pibms_regs2dump.H>
+#include <ekb/hwpf/fapi2/include/return_code_defs.H>
+#include <fapi2.H>
+
+#include <filesystem>
+#include <fstream>
+#include <string>
+#include <vector>
+
+namespace openpower::phal
+{
+namespace dump
+{
+
+using namespace openpower::phal::logging;
+
+void preCollection(struct pdbg_target* proc)
+{
+	struct pdbg_target *fsi, *pib;
+	char path[16];
+	sprintf(path, "/proc%d/fsi", pdbg_target_index(proc));
+	fsi = pdbg_target_from_path(NULL, path);
+	assert(fsi);
+	sprintf(path, "/proc%d/pib", pdbg_target_index(proc));
+	pib = pdbg_target_from_path(NULL, path);
+	assert(pib);
+
+	if ((pdbg_target_probe(fsi) != PDBG_TARGET_ENABLED ||
+	     pdbg_target_probe(pib) != PDBG_TARGET_ENABLED)) {
+		pdbg_target_status_set(proc, PDBG_TARGET_DISABLED);
+	}
+}
+
+int writeDumpFile(char* data, size_t len, std::filesystem::path& dumpPath)
+{
+	int rc = -1;
+	std::ofstream outfile{dumpPath, std::ios::out | std::ios::binary};
+	if (!outfile.good()) {
+		int err = errno;
+		log(level::INFO,
+		    "Failed to open the dump file for writing err=%d", err);
+		return rc;
+	}
+	outfile.exceptions(std::ifstream::failbit | std::ifstream::badbit |
+			   std::ifstream::eofbit);
+	try {
+		outfile.write(data, len);
+	} catch (std::ofstream::failure& oe) {
+		int err = errno;
+		log(level::INFO,
+		    "Failed to write to dump file err=%d error message=%s", err,
+		    oe.what());
+		return rc;
+	}
+	outfile.close();
+	return 0;
+}
+
+int collectSBEDump(uint32_t id, uint8_t failingUnit,
+		   std::filesystem::path& dumpPath)
+{
+
+	int rc = -1;
```

Comment:
Move pdbg related intlization call as part of precollection method , till line 106
Use C++ way ( exception based ) error handling.

Also need initialization on libekb and DEVTREE backend .

These function can define in utils_phal.C /utils_phal.H